### ISSUE TITLE
Rule proposal: detect unused generator calls

### ISSUE THREAD

**Issue by brainkim** [typescript-eslint/typescript-eslint#1068]
Possible documentation for this rule:

# Disallow unused generators (no-unused-generator)
## Rule details
Generator functions are lazy. When called, the execution of the generator suspends at the top of the function until the first time `next` is called on the returned generator object. Calling generator functions without using the result in some way is almost always an error, insofar as you probably expected the code within the function to run eagerly when in actuality it never runs at all.

Examples of *incorrect* code for this rule:
```ts
declare function sideEffect: Generator<number>;
declare function asyncSideEffect: AsyncGenerator<number>;
sideEffect();
asyncSideEffect();
```

Examples of *correct* code for this rule:
```ts
declare function sideEffect: Generator<number>;
declare function asyncSideEffect: AsyncGenerator<number>;
// assignment to a variable
const iter = sideEffect();
const asyncIter = asyncSideEffect();
// calling next immediately
sideEffect().next();
asyncSideEffect().next();
// usage in call expression
Array.from(sideEffect());
// usage in spread expression
const numbers = [...sideEffect()];
// assignment to class property
class View {
  generator = sideEffect();
}

// usage in for loop
for (const number of sideEffect()) {
}

// usage via return or yield
function fn(): Generator<number> {
  return sideEffect();
}

function *gen(): Generator<number> {
  yield *sideEffect();
}
```

**Comment by JoshuaKGoldberg** [comment]
Per discussion in #7749, this has aged away. Closing as out-of-scope: but if common JS patterns shift to using generators more, we'd be happy to take this rule in. Cheers!
